### PR TITLE
Refactor factory

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌

### CODE REVIEW

File: src/Factory.php [hunk 1]
```diff
@@ -91,15 +97,34 @@ public function __construct(
...
      */
     public function create($config)
     {
         if ($this->validate) {
             DefinitionValidator::validate($config);
         }
 
-        /** @psalm-suppress MixedReturnStatement */
-        return $this->dependencyResolver->create($config);
+        if (is_string($config)) {
```
Comment (vjik) [inline src/Factory.php:108]: Add `is_string` to `use` section.

File: src/Factory.php [hunk 2]
```diff
@@ -91,15 +97,34 @@ public function __construct(
...
+            $definition->setReferenceContainer($this->factoryContainer);
+        }
+        try {
+            $container = ($this->container === null || $definition instanceof ReferenceInterface) ? $this->factoryContainer : $this->container;
+            return $definition->resolve($container);
+        } finally {
+            if ($definition instanceof ArrayDefinition) {
+                $definition->setReferenceContainer(null);
+            }
+        }
```
Comment (vjik) [inline src/Factory.php:127]:
```suggestion
return $this->createDefinition($config)->resolve($this->factoryContainer);
```

File: src/Factory.php [hunk 3]
```diff
@@ -19,7 +23,8 @@
  */
 final class Factory
 {
-    private DependencyResolver $dependencyResolver;
+    private ?ContainerInterface $container;
```
Comment (vjik) [inline src/Factory.php:26]:
```suggestion
```

File: src/Factory.php [hunk 4]
```diff
@@ -38,7 +43,8 @@ public function __construct(
         array $definitions = [],
         bool $validate = true
     ) {
-        $this->dependencyResolver = new DependencyResolver($container);
+        $this->container = $container;
```
Comment (vjik) [inline src/Factory.php:46]:
```suggestion
```

File: src/Factory.php [hunk 5]
```diff
@@ -134,4 +159,31 @@ public function setMultiple(array $definitions): void
...
+     *
+     * @throws InvalidConfigException
+     */
+    private function createDefinition($config): DefinitionInterface
+    {
+        $definition = Normalizer::normalize($config);
+
+        if (
+            ($definition instanceof ArrayDefinition) &&
+            $this->factoryContainer->hasDefinition($definition->getClass())
```
Comment (vjik) [inline src/Factory.php:174]: `hasDefinition()` also check `class_exists` and return `true` if not exists definition, but exists class. In this case merge definitions is unneccessary.

File: src/FactoryContainer.php [hunk 6]
```diff
@@ -61,86 +62,54 @@ public function __construct(?ContainerInterface $container)
...
-            return $definition->resolve($container);
-        } finally {
-            if ($definition instanceof ArrayDefinition) {
-                unset($this->creatingIds[$definition->getClass()]);
-                $definition->setReferenceContainer(null);
-            }
-        }
+    public function hasDefinition(string $id): bool
+    {
+        return isset($this->definitions[$id]) || class_exists($id);
```
Comment (vjik) [inline src/FactoryContainer.php:104]: Confused by the name of the method and `class_exists`...

File: src/Factory.php (hunk 1 above)
Comment (yiiliveext) [reply src/Factory.php:110]: Fixed

File: src/Factory.php (hunk 2 above)
Comment (yiiliveext) [reply src/Factory.php:129]: This is wrong, see added tests

File: src/Factory.php (hunk 3 above)
Comment (yiiliveext) [reply src/Factory.php:28]: See above

File: src/Factory.php (hunk 4 above)
Comment (yiiliveext) [reply src/Factory.php:48]: See above

File: src/Factory.php (hunk 5 above)
Comment (yiiliveext) [reply src/Factory.php]: Fixed

File: src/FactoryContainer.php (hunk 6 above)
Comment (yiiliveext) [reply src/FactoryContainer.php]: Fixed

Comment (vjik) [reply src/FactoryContainer.php]: Hm... The essence has not changed: name "hasDefinition", but check not only definition, but also `class_exists`.

File: src/Factory.php (hunk 5 above)
Comment (vjik) [reply src/Factory.php]: Not fixed. If factory not container definition, but class exists, then merge definitions is unneccessary.

File: src/FactoryContainer.php (hunk 6 above)
Comment (yiiliveext) [reply src/FactoryContainer.php]: class_exists - is a method to autoload definitions

File: src/Factory.php (hunk 5 above)
Comment (yiiliveext) [reply src/Factory.php]: This is because Normalizer convert class-string to ArrayDefinition

Comment (vjik) [reply src/Factory.php]: If has not definition, but class exists merge is unneccessary.

File: src/FactoryContainer.php (hunk 6 above)
Comment (vjik) [reply src/FactoryContainer.php]: I think need rename method or remove class_exists.

**Pull request merged**